### ISSUE TITLE
Plan for 2.0

### ISSUE THREAD

**Issue by henryiii** [CLIUtils/CLI11#384]
- Make more parts of the library optional, at least in multiheader mode (postponed)
-  Allow the library to be precompiled to address #194. (still will support header only / single header mode) - postponed to 2.x.
   - [x] New make single header tooling
- [x] Drop deprecated `add_set` templated methods
- [x] Drop deprecated extra true/false argument for capture default on `add_option` #597 
- [x] Drop add_complex (no longer required)
- Allow compilation with exceptions turned off (might put off till 3.0)
- [x] Change Config file output default to TOML

**Comment by sergeyklay** [comment]
Can I ask to consider https://github.com/CLIUtils/CLI11/issues/355 as well as https://github.com/CLIUtils/CLI11/issues/353.

Thank you

**Comment by phlptp** [comment]
- [x] remove some backwards compatibility customizations in `expected` and `type_size`  
- [x] change default settings for integer flags #243
- [x] add support for all containers and other wrapper types 
- [ ] work on help formatters for #355, #353 and a few other issues and purposes 
- [x] add a force function to force an option/flag callback to be called regardless of whether the option was passed or not.

**Comment by henryiii** [comment]
So that we can get something out rather than keep waiting for a large change, I'd like to postpone the pre-compile plans (which is mostly not user facing; that is, users don't have to use it) for 2.x. That way I can focus on the exception part (which will be user-facing), and get a release out sooner rather than (much) later. I have a daughter coming soon, so trying to plan out my limited time!

**Comment by henryiii** [comment]
I might be forced to put off the exception refactoring, since that will further delay 2.0 and as I always tell other people, numbers are cheap - we could do a 3.0 if it requires a backwards incompatible change. And 2.0 would already be useful to get out the door. (plus I accidentally deleted some work I had done on it)

I'll try to get the final removals in (#597 and dropping add_complex) and currently open fixes in, and then if I can't get the error reworking in within a week or so, I'll probably just make a release. @phlptp is this fine with you, or do you have anything else that needs to go in?


```cpp
// Only part of work that didn't get deleted, pasting here
#ifndef CLI11_EXCEPTIONS
#ifdef __EXCEPTIONS
#define CLI11_EXCEPTIONS 1
// GCC and Clang will define the above define
#elif __GNUC__
#define CLI11_EXCEPTIONS 0
#else
#define CLI11_EXCEPTIONS 1
#endif
#endif
```

**Comment by phlptp** [comment]
Nothing major, that can’t wait.  If I fix some of the issues in the next week I will make a PR but don’t wait on me for a release.

**Comment by data-man** [comment]
It would be nice to add a regex validator.

**Comment by henryiii** [comment]
Release should happen fairly soon, probably end of this week. Will need to update contributors, changelog, upcoming -> new features labels.

**Comment by henryiii** [comment]
> It would be nice to add a regex validator.

That would require GCC 4.9+. It's easy to add one yourself, though - and maybe it could be added as an optional include. Those are a bit easier in 2.0.

**Comment by phlptp** [comment]
I think a number of the validators should be split into a separate optional include.  Then we would have a little more freedom for a validator library that wouldn't slow down compilation if someone didn't need them.    At that point adding a regex validator shouldn't be an issue either.  But lets get the 2.0 release out first.

**Comment by phlptp** [comment]
@henryiii  are you changing the minimum compiler requirement on 2.0?

**Comment by henryiii** [comment]
No, it't still 4.8+. CentOS 7 is still an important target (and manylinux1 is 4.8 too). But I'd be okay with an optional file having a higher requirement.

**Comment by henryiii** [comment]
This is now out!

**Comment by kgraney** [comment]
Just to confirm here @henryiii, the exception refactoring didn't happen, right?  Is it still planned?

**Comment by henryiii** [comment]
Yes, it was put off to avoid delaying 2.0 too much.
